### PR TITLE
usm: Replace kprobes with tracepoints

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Converts SOWatcher kprobes into tracepoints.
Unifies prebuilt and runtime compilation functions, as they were the same.

### Motivation

Using tracepoints is encouraged as the ABI is stable and change less over the years, furthermore, unlike kprobes, tracepoints does not suffer from misses.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by val06 — APPROVED** [review]

**Review by usamasaqib — COMMENTED** [review]
This PR does not address the root cause of kprobe misses. The kprobe misses you are likely noticing are not actual misses. These happen because the kernel registers a miss if a page fault occurs during the handling of a kprobe (the kprobe is still executed though). This behavior was removed after 5.14. In `do_sys_open` we are using the `bpf_probe_read_user` helper which is susceptible to page faults.
The new method will suppress these **false** kprobe misses, but it won't solve the problem caused by the use of `bpf_probe_read_user`

A more robust solution could be to read the path once it has been read into kernel memory to avoid failures due to page faults.

**Comment by guyarb** [comment]
> This PR does not address the root cause of kprobe misses. The kprobe misses you are likely noticing are not actual misses. These happen because the kernel registers a miss if a page fault occurs during the handling of a kprobe (the kprobe is still executed though). This behavior was removed after 5.14. In do_sys_open we are using the bpf_probe_read_user helper which is susceptible to page faults.
The new method will suppress these false kprobe misses, but it won't solve the problem caused by the use of bpf_probe_read_user
A more robust solution could be to read the path once it has been read into kernel memory to avoid failures due to page faults.

it already happens in `do_sys_open_helper_enter`

**Review by usamasaqib — COMMENTED** [review]
> it already happens in do_sys_open_helper_enter

`fill_path_safe` uses the same pointers as passed to `bpf_probe_read_user`. So it would also fail due to a page fault.

**Comment by guyarb** [comment]
> > it already happens in do_sys_open_helper_enter
> 
> `fill_path_safe` uses the same pointers as passed to `bpf_probe_read_user`. So it would also fail due to a page fault.

The same implemented in other location of the code
https://github.com/DataDog/datadog-agent/blob/c2d3ed1914483cfc10d4970267ff8393fae421de/pkg/network/ebpf/c/protocols/http/buffer.h#L20-L34

what is the difference?

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+typedef struct {
+    unsigned short common_type;
+    unsigned char common_flags;
+    unsigned char common_preempt_count;
+    int common_pid;
+    long syscall_nr;
+} trace_common;
+
+typedef struct {
+    trace_common unsued;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:24]: typo
```suggestion
    trace_common unused;
```

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    int dfd;
+    char* filename;
+} enter_sys_openat_ctx;
+
+typedef struct {
+    trace_common unsued;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:30]: typo
```suggestion
    trace_common unused;
```

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    char buf[LIB_PATH_MAX_SIZE];
+} lib_path_t;
+
+typedef struct {
+    unsigned short common_type;
+    unsigned char common_flags;
+    unsigned char common_preempt_count;
+    int common_pid;
+    long syscall_nr;
+} trace_common;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:21]: I think using a separate struct can cause padding to be inserted at the end, thus potentially causing alignment issues. I see that CWS always uses a single struct and doesn't embed anything.

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    __u32 len;
+    char buf[LIB_PATH_MAX_SIZE];
+} lib_path_t;
+
+typedef struct {
+    unsigned short common_type;
+    unsigned char common_flags;
+    unsigned char common_preempt_count;
+    int common_pid;
+    long syscall_nr;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:20]: I think this should be `int` based on `/sys/kernel/tracing/events/syscalls/sys_enter_openat/format`: `field:int __syscall_nr;	offset:8;	size:4;	signed:1;`

File: pkg/network/usm/shared_libraries_test.go [hunk 5]
```diff
@@ -486,14 +491,14 @@ func initEBPFProgram(t *testing.T) *ddebpf.PerfHandler {
 		Probes: []*manager.Probe{
 			{
 				ProbeIdentificationPair: manager.ProbeIdentificationPair{
-					EBPFFuncName: "kprobe__" + probe,
+					EBPFFuncName: fmt.Sprintf("tracepoint__syscalls__sys_enter_%s", openatSysCall),
```
Comment (brycekahle) [inline pkg/network/usm/shared_libraries_test.go:494]: I think it makes sense to extract a common function for the enter/exit `fmt.Sprintf`.

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:24]: irrelevant after removing trace_common

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h (hunk 2 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:30]: irrelevant after removing trace_common

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/tls/sowatcher-types.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    __u32 len;
+    char buf[LIB_PATH_MAX_SIZE];
+} lib_path_t;
+
+typedef struct {
+    unsigned short common_type;
+    unsigned char common_flags;
+    unsigned char common_preempt_count;
+    int common_pid;
+    long __syscall_nr;
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:20]: making that change cause the tests to fail, as there is an alignment issue
```
guy@arbitman:~ $ sudo cat /sys/kernel/tracing/events/syscalls/sys_enter_openat2/format
[sudo] password for guy: 
name: sys_enter_openat2
ID: 636
format:
	field:unsigned short common_type;	offset:0;	size:2;	signed:0;
	field:unsigned char common_flags;	offset:2;	size:1;	signed:0;
	field:unsigned char common_preempt_count;	offset:3;	size:1;	signed:0;
	field:int common_pid;	offset:4;	size:4;	signed:1;

	field:int __syscall_nr;	offset:8;	size:4;	signed:1;
	field:int dfd;	offset:16;	size:8;	signed:0;
	field:const char * filename;	offset:24;	size:8;	signed:0;
	field:struct open_how * how;	offset:32;	size:8;	signed:0;
	field:size_t usize;	offset:40;	size:8;	signed:0;
```
see __syscall_nr

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/network/ebpf/c/protocols/tls/sowatcher-types.h:20]: I see, there is an extra `4` bytes of offset after `__syscall_nr`.

**Review by brycekahle — APPROVED** [review]

**Pull request merged**